services/galley/src/Galley/API/Query.hs

```diff
@@ -118,15 +122,55 @@ getConversationRoles zusr cnv = do
   --       be merged with the team roles (if they exist)
   pure $ Public.ConversationRolesList wireConvRoles
 
-getConversationIds :: UserId -> Maybe ConvId -> Maybe (Range 1 1000 Int32) -> Galley (Public.ConversationList ConvId)
-getConversationIds zusr start msize = do
+listConversationIdsUnqualified :: UserId -> Maybe ConvId -> Maybe (Range 1 1000 Int32) -> Galley (Public.ConversationList ConvId)
+listConversationIdsUnqualified zusr start msize = do
   let size = fromMaybe (toRange (Proxy @1000)) msize
   ids <- Data.conversationIdsFrom zusr start size
   pure $
     Public.ConversationList
       (Data.resultSetResult ids)
       (Data.resultSetType ids == Data.ResultSetTruncated)
 
+-- | Lists conversation ids for the logged in user in a paginated way.
+--
+-- Pagination requires an order, in this case the order is defined as:
+--
+-- - First all the local conversations are listed ordered by their id
+--
+-- - After local conversations, remote conversations are listed ordered
+-- - lexicographically by their domain and then by their id.
+listConversationIds :: UserId -> Public.GetPaginatedConversationIds -> Galley Public.ConvIdsPage
+listConversationIds zusr Public.GetPaginatedConversationIds {..} = do
+  localDomain <- viewFederationDomain
+  case gpciPagingState of
+    Just (Public.ConversationPagingState Public.PagingRemotes stateBS) -> remotesOnly (mkState <$> stateBS) (fromRange gpciSize)
+    _ -> localsAndRemotes localDomain (fmap mkState . Public.cpsPagingState =<< gpciPagingState) gpciSize
+  where
+    mkState :: ByteString -> C.PagingState
+    mkState = C.PagingState . LBS.fromStrict
+
+    localsAndRemotes :: Domain -> Maybe C.PagingState -> Range 1 1000 Int32 -> Galley Public.ConvIdsPage
+    localsAndRemotes localDomain pagingState size = do
+      localPage <- pageToConvIdPage Public.PagingLocals . fmap (`Qualified` localDomain) <$> Data.conversationIdsPageFrom zusr pagingState size
+      let remainingSize = fromRange size - fromIntegral (length (Public.pageConvIds localPage))
+      if Public.pageHasMore localPage
+        then pure localPage
+        else do
+          remotePage <- remotesOnly Nothing remainingSize
+          pure $ remotePage {Public.pageConvIds = Public.pageConvIds localPage <> Public.pageConvIds remotePage}
+
+    remotesOnly :: Maybe C.PagingState -> Int32 -> Galley Public.ConvIdsPage
+    remotesOnly pagingState size =
+      pageToConvIdPage Public.PagingRemotes <$> Data.remoteConversationIdsFrom zusr pagingState size
+
+    pageToConvIdPage :: Public.ConversationPagingTable -> Data.PageWithState (Qualified ConvId) -> Public.ConvIdsPage
+    pageToConvIdPage table Data.PageWithState {..} =
+      Public.ConvIdsPage
+        { pageConvIds = pwsResults,
+          pageHasMore = isJust pwsState,
+          pagePagingState = Public.ConversationPagingState table (LBS.toStrict . C.unPagingState <$> pwsState)
```

Comment:
I am wary of string conversions and encodings. I always thought it's rather easy for a bug to slip in. A line like this makes me wonder if it is right. Perhaps it is obvious to you, but would it make sense to pull out (to a block of code) string conversion-related operations with the goal to make it clear what gets converted to what, and in hope it would be easier to review? You had those new encoding and decoding functions elsewhere.

Reply:
I can only see `LBS.toStrict` as string conversion here, rest is just aligning types. I have added explicit field names to make it clearer, please let me know if it is ok.

Reply:
My apologies, here it is just one conversion. Those string conversions elsewhere got me thinking and when I saw this one, I made a fuss out of nothing. I'm sorry, it's all good!